components/p3a/p3a_message.cc

```diff
@@ -0,0 +1,121 @@
+/* Copyright 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/p3a/p3a_message.h"
+
+#include <vector>
+
+#include "base/containers/flat_set.h"
+#include "base/cxx17_backports.h"
+#include "base/logging.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/trace_event/trace_event.h"
+#include "brave/components/brave_prochlo/prochlo_message.pb.h"
+#include "crypto/sha2.h"
+
+namespace brave {
+
+MessageMetainfo::MessageMetainfo() = default;
+MessageMetainfo::~MessageMetainfo() = default;
+
+void GenerateP3AMessage(uint64_t metric_hash,
+                        uint64_t metric_value,
+                        const MessageMetainfo& meta,
+                        brave_pyxis::RawP3AValue* p3a_message) {
+  // TODO(iefremov): - create patch for adding `brave_p3a`
+  // to src/base/trace_event/builtin_categories.h
+  // TRACE_EVENT0("brave_p3a", "GenerateP3AMessage");
+  constexpr size_t kDataLength = 64;
+  uint8_t data[kDataLength] = {0};
+
+  // First byte contains the 4 booleans.
+  const char daily = 1;
+  const char weekly = 0;
+  const char monthly = 2;
+  const char first = 0;
+  data[0] = daily | weekly | monthly | first;
+  uint8_t* ptr = data;
+  ptr++;
+
+  // Find out years of install and survey.
+  base::Time::Exploded exploded;
+  meta.date_of_survey.LocalExplode(&exploded);
+  DCHECK_GE(exploded.year, 999);
+  const std::string yos = base::NumberToString(exploded.year).substr(2, 4);
+  meta.date_of_install.LocalExplode(&exploded);
+  DCHECK_GE(exploded.year, 999);
+  const std::string yoi = base::NumberToString(exploded.year).substr(2, 4);
+
+  const std::string metastring =
+      "," + meta.country_code + "," + meta.platform + "," + meta.version + "," +
```

Comment:
`base::StrCat` would be nice

Reply:
I'll drop the entire function soon, we will upload normal JSONs